### ISSUE TITLE
Questions over one() vs all()

### ISSUE THREAD

**Issue by nicthemighty** [mgonto/restangular#1371]
`Restangular.one(route, id)` works for me without passing in the id (against docs). For example "/business/123/invoice", I can use `Restangular.one("business", 123).one("invoice")` to get the invoice for business 123.

`Element.post(subElement, elementToPost, [queryParams, headers])` works for me by passing in three params (ignoring subElement). For example "/stores/mystore", I can use `Restangular.one("stores", "mystore").post({name:"test"}, {}, {testHeader:"hi"})` to post the new object to "mystore".

Both these examples go against the documentation; is there a better way I should be achieving the results? My understanding of one vs all is that it generates the same URL, but should be used more of a statment of intent around the REST api you're calling - and that the results are handled differently using the collections/elements api?

**Comment by daviesgeek** [comment]
Yes `one` works without passing in the id, but that's not recommended. I believe it will end up appending a second slash to the url when making a request.

> Both these examples go against the documentation; is there a better way I should be achieving the results?

I'm not exactly sure what question you're asking. Yes, both examples work without passing in parameters. That's not "going against" the docs, exactly; you're just leaving out information that should be there. 

> one vs all is that it generates the same URL, but should be used more of a statment of intent around the REST api you're calling

One is to define a resource for a single element; all is for defining the main resource, or a collection (array) of elements. I.e.:

```js

// Defines a single user, with an ID of 1
Restangular.one('users', 1)

// Creates an object for the main resource, i.e., you can get a list of users
var users = Restangular.all('users')

// Gets an array of users, with an object that looks like the first example
users.getList() // [{id: 1}, {id: 2}]
```

**Comment by nicthemighty** [comment]
So I have got really confused with the documentation vs what I need to achieve. I have the following examples where I feel I'm misusing Restangular's API but I'm not sure how to fix it. At the moment, it's mostly being used as a way to build REST calls rather than using the features of Restangularization.

1. `POST /{storeName}/items` - JSON payload is the new item to be created
`Restangular.one("","{storeName}").all("items").post({id:"123"});`
I am using an empty route with .one(), as this is at the base of the url, and {storeName} is not a collection.

2. `GET /{storeName}/items` - returns all the items in the store
`Restangular.one("","{storeName}").all("items").getList();`
I am using an empty route with .one(), as this is at the base of the url, and {storeName} is not a collection.

3. `POST /{storeName}/items/{itemId}` - JSON payload is the item to be updated
`Restangular.one("{storeName}","items").post("{itemId}", {id:"123"});`
I am using .one() to access "items", even though it's a collection - so that I can post payload to the subElement {itemId}. Or I could do `Restangular.one("{storeName}","items").all("{itemId}").post({id:"123"});` but then that treats "items" as an element, and "{itemId}" as a collection.

4. `POST /models/{storeName}` - JSON payload is the new store to be created
`Restangular.all("models").all("{storeName}").post({id:"123"});
I am using .all() on "{storeName"} even though it's an element, so that I can post without a subElement.

5. `GET /models/{storeName}/metadata` - returns the metadata for the store
`Restangular.one("models", "{storeName}").one("", "metadata").get();
I am using an empty route because "metadata" isn't a collection

**Comment by nicthemighty** [comment]
Unless of course, I should concatenating paths?
1. `Restangular.all("{storeName}\items").post({id:"123"});`
2. `Restangular.all("{storeName}\items").get();`
5. `Restangular.one("models", "{storeName}/metadata").get();`

Though that doesn't solve 3 and 4.

**Comment by daviesgeek** [comment]
Yes I think you're misunderstanding how Restangular is meant to be used. I think what you're looking  for is this:

```js

// First, create two Restangular objects: Store and Item
var Store = Restangular.all('models/{storeName}')
var Item = Restangular.all('{storeName}/items')

// 1:
Item.post({id: "123"}).then(function(newItem) {

  // 3:
  newItem.name = "This will be updated"
  newItem.post()
})

// 2:
Item.getList().then(function(items) {
  // items is an array, with each element being a Restangular item with methods to make REST calls
})

// 4:
Store.post({id:"123"}).then(function(newStore) {
  // New store
})

// 5:
// This is just a shorthand method for ease of use. It could be just as easily written like this:
//      Store.get('metadata').then(...)
Store.addRestangularMethod('getMetadata', 'get', 'metadata')
Store.getMetadata().then(function(metadata) {
  // metadata
})
```

**Comment by nicthemighty** [comment]
Thanks for the quick reply; so conceptually - the decision of when to use one() vs all() isn't to do with what the endpoint "means", but what you are trying to do?

So although `GET /models/{storeName}` returns a single object, `var store = Restangular.all('models/{storeName}')` allows me to do `store.post({id:"123"})`. But obviously `store.getList()` would fail because the GET request doesn't return a list?

Basically, I'm really struggling to understand when to use .one() vs .all(), and if the REST api gives you enough information to decide when each should be used?

**Comment by daviesgeek** [comment]
> the decision of when to use one() vs all() isn't to do with what the endpoint "means", but what you are trying to do?

Very simply put, yes. `One` is for a _single_ element, whereas `all` creates a _resource_ (usually a collection of elements).

> `store.getList()` would fail because the GET request doesn't return a list?

If your server isn't returning an array on the GET request, correct.

Does that answer your question?

**Comment by nicthemighty** [comment]
Hi again,

So I guess where I'm really struggling to understand is how to decide whether `one` or `all` is appropriate.

So if I had a REST api with the following endpoints : 

1. `GET /items` - gets all items in the store
2. `POST /items` - create a new item using a payload's contents
3. `GET /items/123` - get an item with id 123
4. `POST /items/123` - update item 123 using payload's contents
5. `GET /items/123/owner` - get the owner for item 123

1. `Restangular.all("items").getList()` is appropriate because I want a list of all items
2. `Restangular.all("items").post()` is appropriate because I want to post a new item to that end point
3. `Restangular.one("items", "123").get()` is appropriate because I want to get item 123

The problem I still have is with 4 and 5
4. `Restangular.all("items").all("123").post()` is the only way in the Restangular API to issue a post to that endpoint? Because if I did .one("items", "123").post() I need a subElement to post to.
5. `Restangular.one("items", "123").all("owner").get([0])` - but then /owner only returns a single element and not a list - so I assume it would cause an error as .all().get() works against a collection?

Basically I'm struggling from the documentation and the conversation above to understand the right way to achieve 4 and 5 based off what each endpoint returns. And what .all() and .one() mean in these two cases.

Sorry if we're going round in circles, just the documentation really doesn't make it clear over these somewhat edge cases.

**Comment by daviesgeek** [comment]
I guess I'm not understanding your use case for Restangular. If you're just making one-off HTTP requests, and you aren't using a full Restangular object, you're better off using `$http`, rather than dealing with the complexities of Restangular.

4:
> is the only way in the Restangular API to issue a post to that endpoint?

Short answer, yes. But again, if you're just making one-off request, use `$http` instead.

5: No, that will work correctly, as `getList` is the only method expecting an array response.

Also, quick question, are you testing any of this code and trying it out with a server? Or just pull it up in a Plunk or something of the like.

**Comment by nicthemighty** [comment]
Hi, I am testing this on a server; I've not had a chance to load up a Plunkr though. For now, I agree that I should really be using $http but I'd like to understand *how* Restangular could be used to solve the problem for the future.

Today, I tried the following as solutions: 

- `POST /items/123` by using `Restangular.all("items").customPOST({id:123}, "123", {}, {});`
- `GET /items/123/owner` by using `Restangular.one("items", "123").customGET("owner", {}, {});`

Is this a good idea? It removes the incorrect usage of .all() or .one() for points 4 & 5, and I assume is why the custom methods were added?

I agree that I should really be using `get().then(function(elem) { elem.field = "newvalue"; elem.save()});` as the correct way to update an item in Restangular for point 4. 

What I still don't understand is how I would access the following endpoint using Restangular (which also is the same issue with point 5): 

`GET /{itemName}` - it's an element, but at the root of the URL so using `.one(route, id)` I would have to pass in "" as the root.. But `.all()` doesn't make sense as it's not a collection.

**Comment by daviesgeek** [comment]
> As it's in the documentation I assume it's the correct way to solve it, is it because of these use cases that custom methods were added?

Both of those examples work.

> `GET /{itemName}`- it's an element, but at the root of the URL

`.one(route, id)` will work but it really isn't the best way to do it.

> `GET /{storeName}/items`

`Restangular.all({storeName} + "/items").getList()` should work like you're using it.
